gpkit/keydict.py

```diff
@@ -0,0 +1,234 @@
+    possible string interpretations of each key.
+
+    Creating a KeyDict:
+    >>>> kd = gpkit.keydict.KeyDict()
+
+    Now kd[x] can be set, where x is any gpkit Variable or VarKey.
+    __getitem__ is such that kd[x] can be accessed using:
+     - x
+     - x.key
+     - x.name (a string)
+     - "x_modelname" (x's name including modelname)
+     - "{x}_{modelname}" (by convention, for latex)
+
+    In addition, if collapse_arrays is True then VarKeys which have a `shape`
+    parameter (indicating they are part of an array) are stored as numpy
+    arrays, and automatically de-indexed when a matching VarKey with a
+    particular `idx` parameter is used as a key.
+
+    By default a KeyDict will regenerate the list of possible key strings
+    for every usage; a KeyDict may instead be "baked" to have a fixed list of
+    keystrings by calling the `bake()` method.
+
+    Note that if a item is set using a key that does not have a `.key`
+    attribute, that key can be set and accessed normally.
+
+    See also: gpkit/tests/t_keydict.py.
+    """
+    collapse_arrays = True
+
+    def __init__(self, *args, **kwargs):
+        "Passes through to dict.__init__ via the `update()` method"
+        # pylint: disable=super-init-not-called
+        self.baked_keystrs = None
+        self.update(*args, **kwargs)
+
+    def update(self, *args, **kwargs):
+        "Iterates through the dictionary created by args and kwargs"
+        for k, v in dict(*args, **kwargs).items():
+            self[k] = v
+
+    def bake(self):
+        "Permanently sets the keystrs of a KeyDict"
+        self.baked_keystrs = None
+        self.baked_keystrs = self.keystrs()
+
+    def keystrs(self):
+        "Generates the strings that may be used as keys for a KeyDict"
+        if self.baked_keystrs:
+            return self.baked_keystrs
+        keystrs = defaultdict(set)
+        for key in self.keys():
+            if hasattr(key, "allstrs"):
+                for keystr in key.allstrs:
+                    keystrs[keystr].add(key)
+        return keystrs
+
+    @classmethod
+    def with_keys(cls, keyset, dictionaries):
+        "Generates a KeyDict from a KeySet and iterable of dictionaries"
+        out = cls()
+        for dictionary in dictionaries:
+            for key, value in dictionary.items():
+                # The keyset filters and converts each dictionary's keys
+                keys = keyset[key]
+                for key in keys:
+                    if not key.idx:
+                        out[key] = value
+                    else:
+                        if not hasattr(value, "shape"):
+                            value = np.array(value)
+                        val_i = value[key.idx]
+                        if is_sweepvar(val_i) or not np.isnan(val_i):
+                            out[key] = val_i
+        return out
+
+    def __contains__(self, key):
+        "In a winding way, figures out if a key is in the KeyDict"
```

Comment:
(it now takes slightly less time to build GasHALE than to solve it, but we seem to also have a results-packing bottleneck because that's taking 24% of solve time)